### ISSUE TITLE
Changing AppId from script before build does not work as expected

### ISSUE THREAD

**Issue by tenkiX** [googlesamples/unity-jar-resolver#365]
### Please fill in the following fields:
Unity editor version: 2019.3.12f1
External Dependency Manager version: 1.2.153
Source you installed EDM (from .unitypackage or Unity Package Manager): tried both ways
Features in External Dependency Manager in use (Android Resolver, iOS Resolver, VersionHandler, etc.): Android resolver, versionhandler
Plugins SDK in use (Firebase, Admob, Facebook, etc.): Firebase Cloud Messaging 6.14 only
Platform you are using the Unity editor on (Mac, Windows, or Linux): Windows
Platform you are targeting (iOS, Android, and/or desktop): Android only
Scripting Runtime (Mono, and/or IL2CPP): IL2CPP (but it does not work on mono as well)

### Please describe the issue here:
When multiple google-services.json files are provided in the Editor, changing the application identifier (package name) and starting the build process via code, the resolver fails to do its job properly and fails to correctly configure cloud messaging (the app starts, but it is linked with the wrong appId for the push notifications). (If the appId is changed from the Editor, it works as expected, the resolver changes everything correctly.)
Enable resolution on build is turned on.

Steps to reproduce:
0. Create new project.
1. Add firebase cloud messaging sdk.
2. Add the following build script to Assets/Editor folder
```
using UnityEditor;

public class Build
{
	[MenuItem("Tools/Build/Debug")]
	public static void BuildDebug()
	{
		PlayerSettings.productName = "debug";
		PlayerSettings.SetApplicationIdentifier(BuildTargetGroup.Android, "com.firebaseExample.debug");
		BuildPipeline.BuildPlayer(new[] { "Assets/Scenes/SampleScene.unity" }, "debug.apk", EditorUserBuildSettings.activeBuildTarget, BuildOptions.None);
	}

	[MenuItem("Tools/Build/Release")]
	public static void BuildRelease()
	{
		PlayerSettings.productName = "release";
		PlayerSettings.SetApplicationIdentifier(BuildTargetGroup.Android, "com.firebaseExample2.release");
		BuildPipeline.BuildPlayer(new[] { "Assets/Scenes/SampleScene.unity" }, "release.apk", EditorUserBuildSettings.activeBuildTarget, BuildOptions.None);
	}

}
```
3. Add google-services.json file(s)
4. Use Tools/Build options from the menu

Please note, that I tried to call:

GooglePlayServices.PlayServicesResolver.Resolve(
			()=> { BuildPipeline.BuildPlayer(new[] { "Assets/Scenes/SampleScene.unity" },"debug.apk", EditorUserBuildSettings.activeBuildTarget, BuildOptions.None); },true);

but the same wrong behaviour observed. Is there any workaround for this? I would need to be able to run the resolver correctly when it is called from a script.

### Please answer the following, if applicable:
What's the issue repro rate? (eg 100%, 1/5 etc)
100%

**Comment by chkuang-g** [comment]
@tenkiX 

I don't think you can dynamically switch bundle id for different google-services.json because Android build relies on `google-services.xml` generated under `Assets/Plugins/Android/Firebase/res/values`, which can only be generated from one of your `google-services.json`.

I don't think Firebase currently supports multiple `google-services.json`.  If you have a strong use case for this, please file a feature support request at https://github.com/firebase/quickstart-unity/issues

Shawn

**Comment by tenkiX** [comment]
Hi @chkuang-g 
https://firebase.google.com/docs/projects/multiprojects
Please check "Support multiple environments in your Android application" section.
The described method works in Unity Editor as expected when appid is switched in the Editor in project settings. (I believe something runs in the background that finds the matching google-services file for the given appId). It must be intended to be supported as:
1. it is documented
2. when you type a wrong appId in unity editor, it prompts you to fix appId and you can select any of your appids defined in all of your google-services files from a dropdown.
But does not work properly when appid is changed from code. However, if you run that build script twice, then only every second build is built correctly. 
I believe that maybe some part of the project resolution runs either asynchronously or after the build, which causes this behavior.

**EXAMPLE**:
Basically, we are doing automated command line builds which calls these functions to build.
what we are experiencing is the following:
0. we set to debug appid in editor.
1. we build with a debug appid config (from command line) - this build is okay, will recieve push notifications from debug config
2. we build with a release appid config (from command line) - this  build will still recieve push notifications from debug config
3. we build AGAIN with a release appid config (still from command line) - this build is okay, will recieve push notifications from release config.

**Comment by master-lincoln** [comment]
I observed a similar problem using only one `google-services.json` file.
After switching the application identifier, the xml generation wasn't triggered.

This is because it's implemented as an asset import post processor which apparently isn't triggered when only changing the app id (Unity 2018.4).
My workaround was to trigger it manually via Reflection:
```
Type type = Type.GetType("Firebase.Editor.GenerateXmlFromGoogleServicesJson, Firebase.Editor");
var method = type.GetMethod("OnBundleIdChanged", BindingFlags.Public | BindingFlags.NonPublic| BindingFlags.Static);
method.Invoke(null, new[] {
	(object)null,
	new PlayServicesResolver.BundleIdChangedEventArgs
	{
		BundleId = newAppId,
		PreviousBundleId = oldAppId
	}
});
```

**Comment by patm1987** [comment]
Some things you could try, let me know how they work:
The file that's generated is in this case is `Assets/Plugins/Android/Firebase/res/values/google-services.xml` (note that in the near future, this may become `Assets/Plugins/Android/Firebase.androidlib/res/values/google-services.xml` to maintain compatibility with Unity 2020.1). You can try explicitly deleting this file to see if it gets regenerated.

You can also try explicitly calling [`AssetDatabase.Refresh()`](https://docs.unity3d.com/ScriptReference/AssetDatabase.Refresh.html) after deleting `google-services.json` (and `google-services.xml` if you do that as well) if this is part of your build process (ex: if your build machine swaps out the `google-services.json` then executes unity with a command line argument to build, make sure this call is in your build function).

If this isn't the case, and you're just dropping it in. Unity should be smart enough to refresh when you gain focus. But it would be interesting to know if right clicking the asset and selecting reimport could help.

As @chkuang-g stated above, there is no current flow in the Unity SDK to have multiple Firebase projects in one Unity project in the same way that the linked iOS and Android tutorial does. There is definitely interest in this, and the best course is to open up an [issue on the Unity SDK page](https://github.com/firebase/quickstart-unity/issues) or contact [Firebase Support directly](https://firebase.google.com/support).

Note: @tenkiX - if you're doing something like swapping out the appid after exporting an Android project from Unity, you can try regenerating just the `google-services.xml` file and dropping this in to your exported Android project. The file that does that ships as [`generate_xml_from_google_services_json.py` (link to C++ SDK version)](https://github.com/firebase/firebase-cpp-sdk/blob/master/generate_xml_from_google_services_json.py).

**Comment by chkuang-g** [comment]
@master-lincoln 

This is an interesting workaround.  

Well, the reason why `PlayServicesResolver.BundleIdChanged` is not triggered immediately is because we are using polling on the main thread to trigger this event.
https://github.com/googlesamples/unity-jar-resolver/blob/825901fa297065d651709fdc34cc5433410c8869/source/AndroidResolver/src/PlayServicesResolver.cs#L1370

Hmm, this is definitely a use-case we did not expect.  I think your workaround should work for this case.  Also, I think we can make `GenerateXmlFromGoogleServicesJson.UpdateJsonWithBundleIdChooserDialog(string bundleId,                                          bool ignoreModificationDate)` public.  This is definitely an undocumented API. lol.

Please give us some time to think about a proper way to support this.  I will mark this as a feature request for now.

**Comment by chkuang-g** [comment]
BTW, I think this is an Firebase feature instead of EDM4U one.  

I created a new issue here.
https://github.com/firebase/quickstart-unity/issues/768

And I will close this one for now for management.  Please up-vote or comment the one in Firebase Quickstart Unity.
